### ISSUE TITLE
[Ecosystem] TorchJD

### ISSUE THREAD

**Issue by PierreQuinton** [pytorch-fdn/ecosystem#66]
### Contact emails

pierre.quinton@gmail.com, valerian.rey@gmail.com

### Project summary

Efficient multi-objective optimization in PyTorch using Jacobian descent and conflict-free updates

### Project description

When optimizing several objectives (losses) at the same time, a common approach is to optimize their average. However, a step that improves the average can still make some individual objectives worse. This happens because the update direction is in conflict with some of the objectives.

If we look at all the gradients together (the Jacobian), we get more information about how the objectives interact. This makes it possible to choose update directions that avoid these conflicts, so that no objective is locally worsened.

This library provides tools to use this information and perform multi-objective optimization in a more careful and consistent way. Compared to a naive implementation, it also improves runtime and reduces memory usage, making it practical for larger problems.

Concretely, TorchJD provides:
- the `autojac` package, wrapping `autograd` to compute Jacobians instead of simple gradients. Its interface is analog to the interface of `autograd`. See for example the [autojac.backward](https://torchjd.org/stable/docs/autojac/backward/) function.
- the `autogram` package, also wrapping `autograd` but at a lower level, to compute the Gramian of the Jacobian (i.e. the Jacobian multiplied by its transpose) as efficiently as possible. See for example the [autogram.Engine](https://torchjd.org/stable/docs/autogram/engine/) documentation.
- the `aggregation` package, to aggregate Jacobians or Gramians into update directions that are beneficial to all objectives at the same time. See for example the [aggregation package](https://torchjd.org/stable/docs/aggregation/) documentation.

### Are there any other projects in the PyTorch Ecosystem similar to yours?  If, yes, what are they?

N/A.

### Project repo URL

https://github.com/SimplexLab/TorchJD

### Additional repos in scope of the application

N/A.

### Project license

MIT License

### GitHub handles of the project maintainer(s)

@ValerianRey , @PierreQuinton 

### Is there a corporate or academic entity backing this project?  If so, please provide the name and URL of the entity.

N/A.

### Website URL

https://torchjd.org

### Documentation

https://torchjd.org

### How do you build and test the project today (continuous integration)?  Please describe.

1. **Checks** (on PRs and daily via cron): A matrix of test configurations runs across Python versions 3.10–3.14, on Ubuntu, macOS, and Windows, with variations in dtype (float32/float64) and dependency sets (full, minimal, lower-bounds). Unit tests run with `pytest -W error` with coverage uploaded to Codecov. We cover all lines of code. Additionally, it builds and doctests the Sphinx documentation, checks link correctness with `lychee`, and enforces code quality via `ty` (type checker) and `pre-commit` hooks (linting/formatting with `ruff`).
2. **Build and Release** (on GitHub Release published): The project uses setuptools as the build backend (configured in pyproject.toml). Builds are produced with `uv build`, which generates both a source distribution (.tar.gz) and a pure Python wheel (.whl). The wheel is platform-independent (py3-none-any) since TorchJD is a pure-Python package with no compiled extensions. On release, a GitHub Actions workflow automatically builds and publishes the package to PyPI using Trusted Publishing (OIDC authentication), so no API tokens are stored as secrets.

### Version of PyTorch

We currently have the lower bound `torch>=2.3.0`, we test the lower bound (so `torch==2.3.0`) and the latest version. We do not test a range of versions in between.

### Components of PyTorch

TorchJD leverages several core PyTorch components:
- `torch.autograd`: We wrap `torch.autograd` to compute full Jacobian matrices via vectorized backward passes in the `autojac` module.
- `torch.func`: We use `torch.func.vjp` and `torch.func.functional_call` to compute per-sample Jacobians and Gramians in the `autogram` module.
- `torch.vmap`: We rely heavily on `torch.vmap` to batch Jacobian computations efficiently, enabling parallel differentiation across multiple loss terms.
- `torch.nn`: We use `Modules` in the package `aggregation` to make our API functional and benefit from hooks.
- AutoGrad graph internals: We use `torch.autograd.graph.GradientEdge`, `Node` to hook into PyTorch's computational graph and implement custom backward transforms that computes some statistics efficiently.

### How long do you expect to maintain the project?

We expect the project to gain traction, as preliminary feedback has been very positive. As the user base grows, we plan to expand the group of maintainers and distribute the maintenance effort more broadly.

The current maintainers will support the project for as long as their expertise remains relevant. In the longer term, the SimplexLab community will take over maintenance and continue to support the project for as long as it remains useful.

### Additional information

As part of this project, we are also exploring ways to further improve efficiency. In particular, we are developing custom PyTorch tensors that could, in the long term, be integrated into PyTorch itself, as they have the potential to benefit the broader ecosystem in terms of performance.

**Comment by ValerianRey** [comment]
@michellearoth Looks like the github-project-automation failed to move this to the backlog 😅 Tagging you in case you missed it!

**Comment by michellearoth** [comment]
@ValerianRey weird! I added to the queue, thanks for flagging!

**Comment by hogepodge** [comment]
@ValerianRey I'll get started on the review tomorrow. Thanks for your patience!

**Comment by hogepodge** [comment]
Title | Description | Status | Notes
-- | -- | -- | --
PyTorch Support | Project must demonstrate benefit to the pytorch community and be validated with the latest versions of PyTorch. | Yes | Version support is not explicit beyond >=2.0
Working CI | Projects must have a demonstrable working CI workflow for all declared architectures | Yes |  
Governance | Projects must have documented technical governance defined and implemented and it must be represented in their Github or Gitlab repository. Example: Review the LF's Minimum Viable Governance framework. | Partial | Closed governance structure, managend entirely by SimplexLab. Proprietary open source structure.
User experience | The value proposition should be made clear for the project and documented in the README of the project repository. | Yes |  
Quick start | Project should have a working and demonstrable quick start. | Yes |  
Documentation | Projects must have clear and comprehensive documentation. | Yes |  
Permissive licensing | Users must be able to utilize ecosystem projects without licensing concerns. e.g. BSD-3, Apache-2 and MIT licenses | Yes |  
Functional Testing | Users should have developed tests submitted as a part of their project, and be confident that the project will work well with the latest releases of PyTorch. | Yes |  
Packages | Projects need to have support for binary installation options (pip/Conda). | Yes |  
Ongoing maintenance | Project maintainers must be committed to supporting and maintaining their projects for the next 12 months at a minimum. Projects must commit to updating their projects to the latest two releases of PyTorch (e.g. versions 2.6, 2.5.1). | Yes |  
Core Maintainers | The project must have a minimum of 2 core maintainers | Yes |  
Contributors | The project should have a minimum of 5 total contributors active in the last 90 days. | Yes | Barely crosses requirement threshold, with core contributors making the bulk of commits, and 2-3 commits from other three contributors.
Commits | The project must have 30 commits in the last 90 days. | Yes |  
Stars | The project must have a minimum of 200 stars on github | Yes | 321 Stars
LICENSE.md | at the root of the repository specifying the terms and conditions for using, distributing, and modifying the software. In addition, you should provide information on the license of any third-party code included in the project. | Yes | MIT License
README.md | welcomes new community members to the project and explains why the project is useful and how to get started. Should include information about release methodology, cadence, and criteria in the README. | Yes |  
CONTRIBUTING.md | explains how to contribute to the project. The file explains the types of contributions needed and how the development process works. | Yes |  
CODEOWNERS | that define individuals or teams responsible for code in a repository; document current project owners and Retired committers. | Yes |  
CODE_OF_CONDUCT.md | sets the ground rules for participants' behavior and helps facilitate a friendly, welcoming environment. By default, projects should leverage the Linux Foundation Code of Conduct unless an alternate Code of Conduct is approved prior. | Yes |

**Comment by hogepodge** [comment]
@PierreQuinton thanks for the submission, I don't see any blockers to advancing the project for a vote.

**Comment by PierreQuinton** [comment]
@hogepodge Thank you very much for the review.

If I may comment on governance, SimplexLab is an open-source community. Currently, it exists as a GitHub organization and a Discord server (with 16 members, 4 of whom are regularly active), and its sole activity is developing and maintaining TorchJD.

The reason we chose a distinct name instead of just calling it "TorchJD" is that we want this community to develop new open-source projects in the future. Our envisioned roadmap includes creating new Tensor subclasses for PyTorch, building video games, hosting online courses, and developing collaborative tools. Because these initiatives go beyond the scope of TorchJD, we established SimplexLab as a larger community. In the future, SimplexLab will likely become a non-profit so that we can manage a formal budget.

**Comment by hogepodge** [comment]
Thanks @PierreQuinton! I appreciate the clarification.

**Comment by michellearoth** [comment]
Approved! 

<img width="715" height="954" alt="Image" src="https://github.com/user-attachments/assets/0c0168c0-97d3-4414-861c-1bff9a791fca" />

**Comment by PierreQuinton** [comment]
Many thanks everyone, we are all very happy.

**Comment by hogepodge** [comment]
Congratulations @PierreQuinton!
